### PR TITLE
Improving autofill background appearance on password inputs

### PR DESCRIPTION
**Pull request by anthony-hull**
### Details
When auto filling any normal field we should see the whole input styled by autofill stylings
When autofilling a password input we should see part of the input styled with a strait border and a space before the visibility toggle button.

### Fixed Issues

$ https://github.com/Expensify/App/issues/7016

### Tests

1. At login screen auto fill email box
2. see the background fills the whole box
3. autofill the password box
4. see that the box is filled like the screenshots

- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

### Tested On

- [x] Web
- [x] Mobile Web
- [ ] Desktop
- [x] iOS
- [x] Android

#### Web
image

image

image
Hasn't regressed on other non secure inputs
#### Mobile Web
image
image

#### Desktop
There isn't any autofill on this

#### iOS
iOS doesn't colour backrounds on autofill. N/A

#### Android
image

Fix border underling for inputs on android:
Screenshot_20220119-160937_New Expensify

Screenshot_20220119-160944_New Expensify

### CODE REVIEW

**Comment by anthony-hull** [comment]
Hey team, I'm struggling to test on iOS. The autofill isn't triggering on the login. I would install 1Password but I don't have a dev account to build for physical devices. Do you have an idea?
Also and ideas on how to test autofill on the desktop client?

**Comment by rushatgabhane** [comment]
> Hey team, I'm struggling to test on iOS. The autofill isn't triggering on the login. I would install 1Password but I don't have a paid iOS dev account to build for physical devices, so I'm stuck on the simulator. Do you have an idea?
Also and ideas on how to test autofill on the desktop client?

On it! Just need some time

**Comment by anthony-hull** [comment]
@Luke9389 do you know why that unit test is failing?

**Comment by rushatgabhane** [comment]
> do you know why that unit test is failing?
It's an issue with OSBotify's GPG key. Not related to the PR

**Comment by rushatgabhane** [comment]
@anthony-hull 
I tried all major password managers and couldn't get them to autofill on desktop. Password managers don't support it yet.

On iOS, I can autofill but the color of TextInput stays the same so nothing to show here

File: src/components/TextInput/BaseTextInput.js
Comment (rushatgabhane) [inline src/components/TextInput/BaseTextInput.js:187]: Should we multiline the styles? It's overflowing the screen.

**Review by rushatgabhane — CHANGES_REQUESTED** [review]
Good work! Some suggestions

**Comment by rushatgabhane** [comment]
@anthony-hull
There's a very small white gap at bottom of TextInput when you fill with a password manager (Lastpass) in Android.
Is this fixable?

Screenshot_20220111-161754 (2)

**Comment by anthony-hull** [comment]
I will

> @anthony-hull There's a very small white gap at bottom of TextInput when you fill with a password manager (Lastpass) on Android. The gap size changes when you hide/show password.
> 
> Is this fixable? Screenshot_20220111-161754 (2)

I'll have a look! Should be able to today. If not tomorrow :)

**Comment by anthony-hull** [comment]
I think this happens when you auto fill into the empty field when it's not using the secure input prop. That is when the password is toggled to visible and then you autofill.  

Autofilling the field in its default state I couldn't reproduce this.

**Comment by rushatgabhane** [comment]
> I think this happens when you auto fill into the empty field when it's not using the secure input prop. That is when the password is toggled to visible and then you autofill.

Oh true, are you able fix this?

**Comment by anthony-hull** [comment]
I've not been able to so far. I've not done much native visual debugging before now. 
I can't see anything different about the properties of the native element between the two states in Flipper.

**Comment by anthony-hull** [comment]
do you have any ideas? or advice on how to debug this?

**Comment by rushatgabhane** [comment]
Hmm you could experiment with padding. Maybe 1px more towards the bottom.

**Comment by anthony-hull** [comment]
I tried padding. Didn't seem to make a difference. I'll have another go tomorrow morning!

**Comment by rushatgabhane** [comment]
Alright, I took a look into this and like @anthony-hull said, this only happens when you autofill when password is visible. 

On second thoughts, to notice this you gotta really zoom and pixel peep. 
I think this is good to merge. 

cc: @Luke9389

**Comment by Luke9389** [comment]
Hmm, let me see if I can crack it. It seems like we're off by the width of the border. 
@shawnborton what do you think about this? We've got a line of white pixels at the bottom of this input (I see it in a few other screenshots too).

**Comment by rushatgabhane** [comment]
I was [wrong](https://github.com/Expensify/App/pull/7099#issuecomment-1012738869). Happens when you use google autofill. And the whitespace is much worse

https://user-images.githubusercontent.com/29673073/149624696-ff63083b-4fc7-4d68-ae4d-233dc2cd60a3.mp4

**Comment by shawnborton** [comment]
I think we should try to fix the 1px gap issue before merging this.

**Comment by rushatgabhane** [comment]
> I think we should try to fix the 1px gap issue before merging this.

@anthony-hull gentle nudge :)

**Comment by anthony-hull** [comment]
I'll have another look today

**Comment by anthony-hull** [comment]
I'll have to look tomorrow instead sorry. 
Lines of thinking:
- Is this introduced by this change or was it there before?
- Is this a subpixel division issue? 
- Does line height influence this?
- Does removing borders completely influence it?

**Comment by anthony-hull** [comment]
I'm seeing this happen on the login box with the latest code on main:

Screenshot_20220118-170935_New Expensify

**Comment by rushatgabhane** [comment]
Do you mean the Onyx error? That's unrelated na?

**Comment by anthony-hull** [comment]
> Do you mean the Onyx error? That's unrelated na?

I mean the white line of missing autofill inside the input

**Comment by rushatgabhane** [comment]
Ah, happens on main too

**Comment by anthony-hull** [comment]
Yeah, both screenshots are taken from main

**Comment by anthony-hull** [comment]
So far no. But I find it interesting that the border radius is not being applied to the autofill background on android.
I wonder if it is a bug with react native. Or we are using styles that are unsupported with autofill on that platform.

**Comment by anthony-hull** [comment]
the mismatch increases with a really high bottom padding. It goes away if padding is 0.
I think I've solved it by increasing bottom the padding to 10 and using a minus -2 margin. 
Would you be happy with this solution? I could pass it in index.android.js for the TextInput component

**Comment by Luke9389** [comment]
If this is happening on main, lets not have that be a blocker for this. Has anyone created an issue?

**Comment by rushatgabhane** [comment]
> Has anyone created an issue?
nope

**Comment by rushatgabhane** [comment]
> It goes away if padding is 0.
I think I've solved it by increasing bottom the padding to 10 and using a minus -2 margin.

Nice! Umm, If I'm not wrong, the main usecase of negative margin is to ignore padding. So this should be okay as long as it doesn't break anything.

**Comment by rushatgabhane** [comment]
> we haven't established that it's happening on Android only. Because we both couldn't get TextInput to be colored on iOS

asked about it on [Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1642535851146700) .

**Comment by anthony-hull** [comment]
That is true. 
I just tried the app on my iPhone installed from the App Store and auto fill worked with 1Password and no highlight showed. Looks to me there isn't a background for auto fill on iOS

**Comment by rushatgabhane** [comment]
Bitwarden, Lastpass and iOS password manager dont't highlight. Okay, probably safe to assume there is no highlighting on iOS 😅

**Comment by anthony-hull** [comment]
great news! I'll update the PR tomorrow morning with test screenshots and the fix for this second issue too!

File: src/components/TextInput/index.android.js
Comment (rushatgabhane) [inline src/components/TextInput/index.android.js:16]: Start the comment with capital letter

**Review by rushatgabhane — COMMENTED** [review]
@Luke9389 looks like the 1px gap issue is for `TextInput` component too. 
It's reproducible on `main` for web. 

It's on inputs for pages like DebitCardPage, but not on phone/email input even though both use the same component.

image

This is unrelated to the current PR and issue. What do you suggest is the best course of action?
Anthony's fix for Android doesn't work here because only some `TextInput`s have the 1px gap issue.

**Comment by anthony-hull** [comment]
Feels too much of a hack to do this to all inputs.
I think we need another issue and to have a root cause rather than a hack to hide it

**Comment by rushatgabhane** [comment]
I agree @anthony-hull

**Comment by Luke9389** [comment]
Yup, I agree that we should open a new issue and solve the gap separately.

File: src/components/TextInput/index.android.js
Comment (rushatgabhane) [reply src/components/TextInput/index.android.js:16]: @anthony-hull approving after this is done 👍

**Comment by Luke9389** [comment]
I think @rushatgabhane should be the one to get the bug report credit for that since he mentioned it first, [here](https://github.com/Expensify/App/pull/7099#issuecomment-1009962247). Are you ok with that @anthony-hull?

**Comment by anthony-hull** [comment]
Yeah that makes sense to me!
Thanks for asking

File: src/components/TextInput/index.android.js
Comment (rushatgabhane) [reply src/components/TextInput/index.android.js]: @anthony-hull waiting on this

Comment (anthony-hull) [reply src/components/TextInput/index.android.js]: this file doesn't exist anymore. I reverted the commit.

**Review by rushatgabhane — APPROVED** [review]
@Luke9389 LGTM! 🎉 I've put the white line issue [on slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1642714377248800) so that it can be created.

**Comment by Luke9389** [comment]
Just wanted to let y'all know that I'll review this tomorrow. Super stacked with stuff to do today  😅

**Comment by rushatgabhane** [comment]
Thanks for letting us know, take your time 😄

**Comment by anthony-hull** [comment]
Cool :)
Hoping no changes. I'm on holiday until Tuesday with no access to a computer

**Comment by anthony-hull** [comment]
I'm back today :)

**Comment by rushatgabhane** [comment]
@Luke9389 bump :)

**Comment by Luke9389** [comment]
Thanks, @rushatgabhane. Sorry, I've been swamped recently :D

**Review by Luke9389 — APPROVED** [review]
Thanks @rushatgabhane and @anthony-hull!

**Pull request merged**
